block-node/archive/src/main/java/org/hiero/block/node/archive/ArchivePlugin.java

```diff
@@ -1,19 +1,278 @@
 // SPDX-License-Identifier: Apache-2.0
 package org.hiero.block.node.archive;
 
+import static org.hiero.block.node.base.BlockFile.blockNumberFormated;
+
+import com.hedera.hapi.block.stream.output.BlockHeader;
+import com.hedera.pbj.runtime.ParseException;
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.time.format.DateTimeFormatter;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.LongStream;
+import org.hiero.block.node.base.s3.S3Client;
+import org.hiero.block.node.spi.BlockNodeContext;
 import org.hiero.block.node.spi.BlockNodePlugin;
+import org.hiero.block.node.spi.ServiceBuilder;
+import org.hiero.block.node.spi.blockmessaging.BlockNotificationHandler;
+import org.hiero.block.node.spi.blockmessaging.PersistedNotification;
+import org.hiero.block.node.spi.historicalblocks.BlockAccessor.Format;
+import org.hiero.hapi.block.node.BlockUnparsed;
 
 /**
  * This a block node plugin that stores verified blocks in a cloud archive for disaster recovery and backup. It will
  * archive in batches as storing billions of small files in the cloud is non-ideal and expensive. Archive style cloud
  * storage has minimum file sizes and per file costs. So batches of compressed blocks will be more optimal. It will
- * watch persisted block notifications for when the next batch of blocks is available to be archived. It will then
- * fetch those blocks from persistence plugins and upload to the archive.
+ * watch persisted block notifications for when the next batch of blocks can be archived. It will then fetch those
+ * blocks from persistence plugins and upload to the archive.
  */
-public class ArchivePlugin implements BlockNodePlugin {
-    // keep track of the last block number archived
-    // have configuration for the number of blocks to archive at a time
-    // on init register as a block notification handler, listen to persisted notifications
-    // on persisted notification, check if the whole batch size of blocks from last block archived is available
-    // if so, fetch them from the persistence plugin, compress and upload to the archive bucket
+public class ArchivePlugin implements BlockNodePlugin, BlockNotificationHandler {
```

Comment:
As structured the `ArchivePlugin` is focused on the cloud bucket archiving.
Should this class rather be more general and the plugin support integration with local archiving and cloud archiving in a staged fashion based on the enabled services?
For example we already have the ZipBlockArchive under the historical block provider, should that run creating local archive files and then the cloud archive service uploads the already created archived files.
In this way the cloud logic is simpler as it's only looking for the compressed archive files and whether it's uploaded the latest, rather than having to track and calculate how many blocks should go into an archive file which is already done.

Reply:
Synced with @jasperpotts and @jsync-swirlds 
This plugin is indeed intentionally focused on cloud archive into S3 API compatible cloud providers.
This is DR focused thus supporting the different approaches in archival file type i.e. tar vs zip, as it allows for more stream like file creation, also background parallel building (expected to be low resource but needs consideration of large chunk size) and other items.